### PR TITLE
test(domain): add DumpTablesJSON method

### PR DESCRIPTION
**Pull request by barrettj12**
The `internal/database/testing` package already has a `DumpTable` method for debugging, which outputs an entire database table in tabular format. However, this tabular format is unsuitable for tables with many columns, or with particularly long values. In this case, a "JSON-like" format makes more sense.

This PR adds a modified version of the `DumpTable` method, which outputs the specified tables in a "JSON-like" format. The output is not and should not be treated as valid JSON.

Under the hood, I have changed both methods to use a common `dumpTable` method, which accepts a `formatter` interface that defines the output format. There are implementations of `formatter` for tabular and "JSON-like" formats, and it would be easy to add more formats in the future if desired.

Also, both methods now accept multiple table names, for convenience.

A reminder that these methods are intended **only** for debugging during development. They should not be used in any tests committed to the repo, and certainly not in production.

## Example

```go
databasetesting.DumpTablesJSON(c, s.DB(), "model", "user", "controller_config", "foobar")
```

<details>
<summary>Output</summary>

```
***** TABLE "model" *****
{
  "activated": true,
  "cloud_credential_uuid": "88a5a93b-381d-46a8-8a77-7518aca2bc71",
  "cloud_region_uuid": "bb5d8996-daf0-4315-8760-3e76e05f5586",
  "cloud_uuid": "7d81fc41-4f16-40e3-87a7-68ad3448c71d",
  "life_id": 0,
  "model_type_id": 0,
  "name": "my-test-model",
  "owner_uuid": "0755f8fa-a448-4888-811b-70137ebd2194",
  "uuid": "0a8ddcf5-08be-45b6-8aeb-81a2e455bf85"
}
{
  "activated": true,
  "cloud_credential_uuid": "88a5a93b-381d-46a8-8a77-7518aca2bc71",
  "cloud_region_uuid": "bb5d8996-daf0-4315-8760-3e76e05f5586",
  "cloud_uuid": "7d81fc41-4f16-40e3-87a7-68ad3448c71d",
  "life_id": 0,
  "model_type_id": 0,
  "name": "listtest1",
  "owner_uuid": "0755f8fa-a448-4888-811b-70137ebd2194",
  "uuid": "e7f9538b-7935-4671-8194-ed652e87d073"
}
{
  "activated": true,
  "cloud_credential_uuid": "88a5a93b-381d-46a8-8a77-7518aca2bc71",
  "cloud_region_uuid": "bb5d8996-daf0-4315-8760-3e76e05f5586",
  "cloud_uuid": "7d81fc41-4f16-40e3-87a7-68ad3448c71d",
  "life_id": 0,
  "model_type_id": 0,
  "name": "listtest2",
  "owner_uuid": "0755f8fa-a448-4888-811b-70137ebd2194",
  "uuid": "3d9c0759-3fa4-4e26-8ec2-c8177a51aede"
}

***** TABLE "user" *****
{
  "created_at": "2024-08-02T14:40:11.636835417+12:00",
  "created_by_uuid": "0755f8fa-a448-4888-811b-70137ebd2194",
  "display_name": "test-user",
  "external": false,
  "name": "test-user",
  "removed": false,
  "uuid": "0755f8fa-a448-4888-811b-70137ebd2194"
}

***** TABLE "controller_config" *****

no such table: foobar
```
</details>

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- ~[ ] Go unit tests, with comments saying what you're testing~
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

### CODE REVIEW

**Review by SimonRichardson — COMMENTED** [review]
This already exists, see: https://github.com/juju/juju/blob/main/internal/database/testing/database.go#L19

File: internal/database/testing/database.go [hunk 1]
```diff
@@ -51,3 +52,60 @@ func DumpTable(c *gc.C, db *sql.DB, table string) {
...
+}
+
+func printJSON(c *gc.C, table *tableData) {
+	fmt.Printf("***** TABLE %q *****\n", table.tableName)
+	stdout := json.NewEncoder(os.Stdout)
+
+	for _, row := range table.rows {
+		jsonRow := map[string]string{}
+		for i, columnName := range table.columnNames {
+			jsonRow[columnName] = fmt.Sprintf("%s", *row[i].(*[]byte))
```
Comment (tlm) [inline internal/database/testing/database.go:103]: This is pretty reckless just taking a random pointer to some data and turning it into a bag of bytes for a string. I don't think this is going to work how you expect it an also produce some pretty invalid json. Why not just let the json encoder assume the type and output the correct json type for the val?

File: internal/database/testing/database.go [hunk 2]
```diff
@@ -51,3 +52,60 @@ func DumpTable(c *gc.C, db *sql.DB, table string) {
...
+// tableData is an intermediate representation of an arbitrary table, ready to
+// be processed for printing.
+type tableData struct {
+	tableName   string
+	columnNames []string
+	rows        [][]any
+}
+
+func printJSON(c *gc.C, table *tableData) {
+	fmt.Printf("***** TABLE %q *****\n", table.tableName)
```
Comment (tlm) [inline internal/database/testing/database.go:97]: You are not really printing JSON when this line is in here. I think if we are going to say this is doing json make it all json.

Turn the struct into an object of table name keys that are arrays of objects. That way it is all valid from top to bottom.

**Review by tlm — CHANGES_REQUESTED** [review]
Thanks @barrettj12

File: internal/database/testing/database.go [hunk 3]
```diff
@@ -51,3 +52,60 @@ func DumpTable(c *gc.C, db *sql.DB, table string) {
 	writer.Flush()
 	fmt.Fprintln(os.Stdout)
 }
+
+// DumpTablesJSON prints the named tables to stdout in a JSON format. This is
+// intended as a debugging tool for state tests. It is not intended for use in
+// production code.
+func DumpTablesJSON(c *gc.C, db *sql.DB, tableNames ...string) {
```
Comment (SimonRichardson) [inline internal/database/testing/database.go:59]: I don't get why we have both, one that's JSON and one that is tabwriter. I have a preference for the latter as it's much easier to grok at a glance.

File: internal/database/testing/database.go (hunk 1 above)
Comment (barrettj12) [reply internal/database/testing/database.go:103]: Good point

File: internal/database/testing/database.go (hunk 3 above)
Comment (barrettj12) [reply internal/database/testing/database.go:59]: This is how the output of `DumpTable` displays in my terminal. Hopefully it illustrates why I'm not a fan.
```
uuid                                    activated       cloud_uuid   c
loud_region_uuid                        cloud_credential_uuid        m
odel_type_id    life_id name            owner_uuid                   
2885297f-46ee-48ca-84ea-28589de8f18a    true            1bc930de-8161-
4a2f-8a95-d779d01cbe60  aeb2a28c-6248-4931-88a0-94d79bfc415d    ecfece
dc-717b-4bba-8be6-408c6344b690  0               0       my-test-model4
f821fc6-337d-4536-875a-35e6dbbcb98c     974c3956-31b6-4031-8c73-bf4c18
588eab  true    1bc930de-8161-4a2f-8a95-d779d01cbe60    aeb2a28c-6248-
4931-88a0-94d79bfc415d  ecfecedc-717b-4bba-8be6-408c6344b690    0    0
listtest1       4f821fc6-337d-4536-875a-35e6dbbcb98c    b13a4188-3b62-
4e63-880c-2dedb5d1e1fe  true    1bc930de-8161-4a2f-8a95-d779d01cbe60 a
eb2a28c-6248-4931-88a0-94d79bfc415d     ecfecedc-717b-4bba-8be6-408c63
44b690  0       0       listtest2       4f821fc6-337d-4536-875a-35e6db
bcb98c
```

> I have a preference for the latter

Great, and I have a preference for the JSON. So surely there should be some room for personal preference in these utilities?

File: internal/database/testing/database.go (hunk 2 above)
Comment (barrettj12) [reply internal/database/testing/database.go:97]: This is solely for debugging. No one should be treating the output of this function as valid JSON, and I don't want to encourage that either. The fact that it resembles JSON is coincidental. I could have gone with any format that displays the data in this way.

I don't think your proposal would achieve much other than make the output less readable. If your problem is calling the function "JSON" when it's not actually valid JSON, then sure, I'm open to other name suggestions.

File: internal/database/testing/database.go (hunk 3 above)
Comment (SimonRichardson) [reply internal/database/testing/database.go:59]: Then pass in a formatter that allows you to express that.

```go
func DumpTablesJSON() { ... }
func DumpTables() { ... }

func dumpTables(..., formatted dumpTablesFormatter)
```

File: internal/database/testing/database.go [hunk 4]
```diff
@@ -14,102 +14,138 @@ import (
...
 
+func (t *tabularFormatter) AddRow(vals []any) {
+	for _, val := range vals {
+		fmt.Fprintf(t.tw, "%v\t", *val.(*any))
+	}
+}
+
+func (t *tabularFormatter) Flush() {
+	fmt.Fprintln(os.Stdout)
+	fmt.Printf("***** TABLE %q *****\n", t.tableName)
```
Comment (SimonRichardson) [inline internal/database/testing/database.go:104]:
```suggestion
	fmt.Fprintf(os.Stdout, "Table: %q\n", t.tableName)
	fmt.Fprintf(os.Stdout, strings.Repeat("-", len(t.tableName) + 7))
```

You don't need to scream about it.

File: internal/database/testing/database.go [hunk 5]
```diff
@@ -14,102 +14,138 @@ import (
...
+func (j *jsonFormatter) SetColumns(cols []string) {
+	j.columnNames = cols
+}
+
+func (j *jsonFormatter) AddRow(row []any) {
+	j.rows = append(j.rows, row)
+}
+
+func (j *jsonFormatter) Flush() {
+	fmt.Printf("***** TABLE %q *****\n", j.tableName)
```
Comment (SimonRichardson) [inline internal/database/testing/database.go:137]: If you're going to keep with JSON, keep with JSON.

**Pull request closed without merging**